Extend to paragraphs

pewgeuges opened the pull request:
Hi Karolína,

I’ve repurposed your great plugin so it adds links to paragraphs instead of headings.

The reasons are:
1. I was looking for a solution to get Douglas Engelbart’s Purple Numbers without the downside of a numbering;
1. I’m using Table of Contents Plus, that already adds anchors to headings, and extended it to make those links grabbable from the headings.

Many thanks, you helped save the WordPress tutorial I’m working on!

pewgeuges committed:
Update class-add-anchor-links.php

repurposed for paragraphs

pewgeuges committed:
Update add-anchor-links.css

repurposed for paragraphs

pewgeuges committed:
Update add-anchor-links.css

added header, refactored code for transparency

pewgeuges committed:
Update class-add-anchor-links.php

updated header, comments, variables to reflect the new focus

pewgeuges committed:
Update class-add-anchor-links.php

fixed added tabs

vyskoczilova commented:
Hi @pewgeuges This is cool, glad it helped you solve your problem. 

Thanks for PR, but I can't merge it in this form - the new changes shouldn't modify the old functionality. I would suggest adding a setting where you can check that you want to add anchors to the paragraphs as well (or only). What do you think? Could you modify your code? 
If not, I'll probably do it later myself, but currently, I'm quite busy with Christmas-related customer requests ;).

pewgeuges committed:
Update README.md

updated readme according to the new purpose

pewgeuges committed:
Update README.md

pewgeuges committed:
Update README.md

pewgeuges commented:
Hi @vyskoczilova,

Thank you for your reply and review. 

Indeed I see I only thought at solving a problem and letting you know, and only later noticed like you that the two should be combined and optional.

Yes I shall add more settings, the way you designed, and perhaps a choice for the headings: link in icon like on GitHub readme.md as you are doing, link all over the heading like on docs.github.com, or no link because TOC already cares.

As hinted in the lately added README.md, for paragraphs the icon way is always the best fit.

I feel rushed in this issue as the WP method should be ready well ahead of Christmas. Stay tuned 😷

pewgeuges committed:
Update README.md

pewgeuges committed:
Update class-add-anchor-links.php

I got stuck with the WP settings API, and the whole thing is behind the spec. So I’m throwing the code as-is and wait for the time to sit down and look how to properly extend it.

Duplicate disambiguation would be needed too.

pewgeuges committed:
Update add-anchor-links.php

not functional!

pewgeuges committed:
Update settings.php

Settings page display OK
checkboxes functional

Backend not working!

pewgeuges committed:
Add files via upload

just for the record

pewgeuges commented:
I’m sorry, @vyskoczilova, I got stuck and just threw the code for the record before sitting down and looking how to fix the mess I’ve made of it.

pewgeuges committed:
Delete settings-detail.png

pewgeuges committed:
Add files via upload

for the record

pewgeuges committed:
Update README.md

account for using an older file for now

pewgeuges committed:
Update class-add-anchor-links.php

meanwhile reverted to paragraphs only

pewgeuges committed:
Update README.md

work is in progress, while 2 files are usable as provisionally intended

pewgeuges committed:
Update README.md

pewgeuges committed:
Update class-add-anchor-links.php

better regex (simpler is better)

pewgeuges committed:
Update README.md

account for better regex

pewgeuges committed:
Update class-add-anchor-links.php

recovered improved comments

pewgeuges committed:
Update class-add-anchor-links.php

add licence in file header

pewgeuges committed:
Update settings.php

pewgeuges committed:
Update add-anchor-links.css

pewgeuges committed:
Update class-add-anchor-links.php

pewgeuges committed:
Update add-anchor-links.php

pewgeuges committed:
Update README.md

pewgeuges committed:
Update README.md

pewgeuges commented:
Hi Karolína,

It now works as you designed except that I don’t get the setting to work properly, but I’m confident that this will take you only a few minutes to fix, so I’ve rushed to update this PR as it’s now fit for regular release. The plugin’s default behavior won’t be changed except for the new scroll offset.

Thanks for your patience and sorry for delay.

pewgeuges commented:
Just see that this is incomplete, the changed assets/css/add-anchor-links.css and admin/settings.php are not reflected above.

I’ll try to do a new PR now.

pewgeuges commented:
Hope this gets through to you, not only the files listed above:
![Screenshot from 2021-04-18 17-45-35](https://user-images.githubusercontent.com/73141620/115151670-36225b00-a06e-11eb-8946-e7cee3146e1f.png)

pewgeuges commented:
The bug seems to be in the data structure of WordPress default settings pages. When a box is unchecked, it disappears from the DB. And when it is not defined, the setting falls back to its default value. So I’ve changed the default for 'headings' to `false`, and the box can now be checked or unchecked. When all boxes are checked, the DB row is full of keys; when unchecked, it’s empty. That of course cannot work.

So what I’ll do is just re-test with all defaults to `false`. Then either there is a need to change for another dashboard, or it’s acceptable there’s a need to first check a box before getting the plugin to add links. Glad to see you have set up a prompt.

pewgeuges committed:
Add files via upload

pewgeuges committed:
Delete add-anchor-links.min.css

pewgeuges committed:
Suggest using a minified stylesheet for production

Although the stylesheet is lightweight, it might be minified.

pewgeuges committed:
Correct default to false because of data structure

pewgeuges committed:
Working loose instead of failing strict comparison

pewgeuges committed:
Update class-add-anchor-links.php

pewgeuges commented:
It’s now fully functional! :rocket:

(Now the readme needs updating.)

pewgeuges committed:
Managed to get comparisons strict with '1'

pewgeuges committed:
Restore full backcompat

pewgeuges committed:
Restore full backcompat.

pewgeuges committed:
Restore full backcompat.

pewgeuges commented:
There is a way to restore full backwards compatibility, by inverting the meaning of the headings checkbox and using long descriptive labels.

pewgeuges commented:
I’m sorry I forgot to rebase my fork, but the changes made yesterday are actually based on current v1.0.4 installed locally. Also cloned vyskoczilova/add-anchor-links and diffed it against the distribution, to assess that the distributed subpackage v1.0.4 is up-to-date. Then started editing but when updating my fork I forgot to ensure traceability.

This is only a last attempt to complete tasks started last year. Now I need to move on completing (offline) another project. I’m relieved and happy to have been given the chance to provide you with this working solution close to what you outlined above.

vyskoczilova commented:
Hi @pewgeuges! Thanks for your effort! I'll have look at it in beginning of  May (I'm finishing a big project) and have a look if I can merge it or do some fix. I'll let you know.
Good luck with your offline project!
Karolina

P.S. For the next time, I would really appreciate not changing any code (formating, structure) you haven't changed.

pewgeuges committed:
Reflect state after debugging the new settings

pewgeuges committed:
Correct meaning.

pewgeuges committed:
Remove (unupdated) last-modified date.

pewgeuges commented:
@vyskoczilova 

Thanks for your feedback, just saw it. But:

> P.S. For the next time, I would really appreciate not changing any code (formating, structure) you haven't changed.

I’m baffled, sorry about this mishap. Just checked the changed-files view. Seems to include your own updates and pretends it was me. Then diffed my edited copy of your plugin against my fresh clone of your GitHub repo, as I did before editing two days ago (https://github.com/vyskoczilova/add-anchor-links/pull/2#issuecomment-822369617 above), and do not see all the changes highlighted by GitHub, only as in the attached file
[add-anchor-linksGH_add-anchor-links.txt](https://github.com/vyskoczilova/add-anchor-links/files/6343323/add-anchor-linksGH_add-anchor-links.txt)

Since this PR has become unusable due to being based on an outdated fork that I failed to rebase two days ago, I’ll probably need to refer to it in a new PR based on a fresh fork. For traceability I’ll use the same edited files as they are now to upgrade the on-coming fork in one go and open a new PR in the wake.

Apologies for the trouble and process garbage I’ve caused while setting up this PR.

P.S.: Okay, just [learned](https://github.community/t/creating-a-second-fork-on-a-repository-from-same-user/2729) that I can’t have more than one fork, I’ll use branches then.

pewgeuges commented:
This PR is now superseded by https://github.com/vyskoczilova/add-anchor-links/pull/4
